library/Vanilla/Embeds/TwitchEmbed.php

```diff
@@ -0,0 +1,209 @@
+<?php
+/**
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
+ */
+
+namespace Vanilla\Embeds;
+
+use Exception;
+
+/**
+ * Class for parsing twitch URLs, fetching twitch OEmbed data, and rendering server side HTML for a twitch video embed
+ */
+class TwitchEmbed extends VideoEmbed {
+
+    const DEFAULT_HEIGHT = 300;
+
+    const DEFAULT_WIDTH = 400;
+
+    /** @inheritdoc */
+    protected $domains = ['www.twitch.tv', 'clips.twitch.tv'];
+
+    /** @var string */
+    private $urlType;
+
+    public function __construct()
+    {
+        parent::__construct('twitch', 'video');
+        $this->urlType = '';
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function matchUrl(string $url) {
+        $data = [];
+        $oembedData =[];
+        $videoID = $this->getUrlInformation($url);
+
+        if (!$videoID) {
+            throw new Exception('Unable to find Twitch Post', 400);
+        }
+
+        if ($this->isNetworkEnabled()) {
+            $oembedData = $this->oembed("https://api.twitch.tv/v4/oembed?url=" . urlencode($url));
+            if (!empty($oembedData)) {
+                $data = $this->normalizeOembed($oembedData);
+            }
+        }
+
+        $queryInfo = $this->getQueryInformation($url) ?? '';
+        $embedUrl = $this->getEmbedUrl($videoID, $queryInfo);
+        if (!$embedUrl) {
+            throw new Exception('Unable to find Twitch Post', 400);
+        }
+        $data['attributes'] = $data['attributes'] ?? [];
+        $data['attributes']['videoID'] = $videoID;
+        $data['attributes']['embedUrl'] = $embedUrl;
+
+        return $data;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function renderData(array $data): string {
+
+        $convertedHeight = (int)$data['height'];
+        $convertedWidth  = (int)$data['width'];
+
+        $attributes = $data['attributes'] ?? [];
+        $embedUrl = $attributes['embedUrl'] ?? '';
+        $height = ($convertedHeight) ? $convertedHeight : self::DEFAULT_HEIGHT;
+        $width = ($convertedWidth) ? $convertedWidth : self::DEFAULT_WIDTH;
+        $name = $data['name'] ?? '';
+        $photoURL = $data['photoUrl'] ?? '';
+
+        $result = $this->videoCode($embedUrl, $name, $photoURL, $width, $height);
+        return $result;
+    }
+
+    /**
+     * Assigns the link type and retrieves the video id.
+     *
+     * @param string $url The posted url.
+     *
```

Comment:
Vanilla's doc tags are usually grouped together with no empty lines between them. I'm seeing this a lot in this PR, though.